track.go

```diff
@@ -1,76 +1,173 @@
+
+	id          string
+	payloadType uint8
+	kind        RTPCodecType
+	label       string
+	ssrc        uint32
+	codec       *RTPCodec
+
+	packetizer rtp.Packetizer
+	receiver   *RTPReceiver
+	senders    []*RTPSender
 }
 
-// NewRawRTPTrack initializes a new *Track configured to accept raw *rtp.Packet
-//
-// NB: If the source RTP stream is being broadcast to multiple tracks, each track
-// must receive its own copies of the source packets in order to avoid packet corruption.
-func NewRawRTPTrack(payloadType uint8, ssrc uint32, id, label string, codec *RTPCodec) (*Track, error) {
-	if ssrc == 0 {
-		return nil, errors.New("SSRC supplied to NewRawRTPTrack() must be non-zero")
+// ID gets the ID of the track
+func (t *Track) ID() string {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.id
+}
+
+// PayloadType gets the PayloadType of the track
+func (t *Track) PayloadType() uint8 {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.payloadType
+}
+
+// Kind gets the Kind of the track
+func (t *Track) Kind() RTPCodecType {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.kind
+}
+
+// Label gets the Label of the track
+func (t *Track) Label() string {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.label
+}
+
+// SSRC gets the SSRC of the track
+func (t *Track) SSRC() uint32 {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.ssrc
+}
+
+// Codec gets the Codec of the track
+func (t *Track) Codec() *RTPCodec {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.codec
+}
+
+// Read reads data from the track. If this is a local track this will error
+func (t *Track) Read(b []byte) (n int, err error) {
+	t.mu.RLock()
+	if len(t.senders) != 0 {
+		t.mu.RUnlock()
+		return 0, fmt.Errorf("this is a local track and must not be read from")
 	}
+	r := t.receiver
+	t.mu.RUnlock()
 
-	return &Track{
-		isRawRTP: true,
-
-		ID:          id,
-		PayloadType: payloadType,
-		Kind:        codec.Type,
-		Label:       label,
-		SSRC:        ssrc,
-		Codec:       codec,
-	}, nil
+	return r.readRTP(b)
+}
+
+// ReadRTP is a convenience method that wraps Read and unmarshals for you
+func (t *Track) ReadRTP(b []byte) (*rtp.Packet, error) {
+	i, err := t.Read(b)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &rtp.Packet{}
+	if err := r.Unmarshal(b[:i]); err != nil {
+		return nil, err
+	}
+	return r, nil
 }
 
-// NewSampleTrack initializes a new *Track configured to accept media.Sample
-func NewSampleTrack(payloadType uint8, id, label string, codec *RTPCodec) (*Track, error) {
-	if codec == nil {
-		return nil, errors.New("codec supplied to NewSampleTrack() must not be nil")
+// Write writes data to the track. If this is a remote track this will error
+func (t *Track) Write(b []byte) (n int, err error) {
+	t.mu.RLock()
+	if t.receiver != nil {
+		t.mu.RUnlock()
+		return 0, fmt.Errorf("this is a remote track and must not be written to")
+	}
+	senders := t.senders
+	t.mu.RUnlock()
+
+	for _, s := range senders {
+		if _, err := s.sendRTP(append([]byte{}, b...)); err != nil {
```

Comment:
TODO -- Should we always copy when write is called?

Reply:
It's least-surprising. I'd vote for keeping it this way for now and exploring performance optimizations later.

Reply:
`io.Writer: Write must not modify the slice data, even temporarily`. Same is true when you write to a `net.Conn` and other stuff like that. Not always true when dealing with pions which should be fixed.

It's better this way because it's on the writer to perform the copy when it knows it will modify the data. Otherwise you have cases like there where I have to assume the underlying implementation will modify my data.

Reply:
@kixelated this is currently the behavior, so I don't think anything to do here.

It would be nice if Go had move semantics, or some official way to pass ownership of a buffer.